### ISSUE TITLE
[Performance] Inference speed discrepancy between two r200 models on CPU

### ISSUE THREAD

**Issue by cyrusbehr** [microsoft/onnxruntime#15630]
### Describe the issue

I am experiencing a large discrepancy in the CPU inference time between two ResNet 200 models. Both models were generated using pytorch 1.12.1, and exported to onnx using onnx v1.11. Upon inspecting the models using netron, they both appear to be identical. I am running inference on an i9 11th gen, Ubuntu 20.04, using onnxruntime python API v1.14.1.

The first model was not actually trained, the weight were simply initialized. We will call this model `r200_dummy.onnx`. The second model has actually been trained, and we will call this `r200_trained.onnx`.

I have written a script to benchmark the runtime performance on CPU for the two models at different optimization levels on my unloaded machine. The code is as follows:

```python
import onnxruntime as ort
import numpy as np
from time import time


def benchmark_model(model_path):
    sess_opt = ort.SessionOptions()
    sess_opt.intra_op_num_threads = 8
    sess_opt.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
    # sess_opt.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED
    # sess_opt.graph_optimization_level = ort.GraphOptimizationLevel.ORT_DISABLE_ALL

    ort_session = ort.InferenceSession(model_path, sess_opt, providers=['CPUExecutionProvider'])
    output_names = [ort_session.get_outputs()[0].name]
    input_name = ort_session.get_inputs()[0].name

    # Run on dummy input
    img = np.ones((1,3,112,112)).astype('float32')

    # Run 5 times to warm things up
    for i in range(5):
        tp = ort_session.run(None, {input_name:img})

    # Benchmark
    num_iterations = 100
    start = time()
    for i in range(num_iterations):
        op = ort_session.run(None, {input_name:img})

    end = time()
    return (end-start)/num_iterations * 1000


time_1 = benchmark_model("./r200_dummy.onnx")
print("R200 dummy avg inference time:", time_1, "ms")

time_2 = benchmark_model("./r200_trained.onnx")
print("R200 trained avg inference time:", time_2, "ms")

```

Here are the inference results at the various optimization levels: 
**ORT_DISABLE_ALL**:
- R200 dummy avg inference time: 113.25390577316284 ms
- R200 trained avg inference time: 158.7948751449585 ms

**ORT_ENABLE_EXTENDED**:
- R200 dummy avg inference time: 111.59055709838867 ms
- R200 trained avg inference time: 157.80184268951416 ms

**ORT_ENABLE_ALL**:
- R200 dummy avg inference time: 63.053832054138184 ms
- R200 trained avg inference time: 283.2158946990967 ms

As can be seen, there is a large discrepancy between the two model inference times. Increasing the optimization level to the max actually slows things down for the trained model. I understand why there might be a large difference when the highest level of optimizations are enabled, as some of the weight layers for the dummy model are all 0s or 1s so entire operations may be able to be ignored, but I don't understand why there is such a large difference between the two when all optimizations are disabled.

To add to this, and what actually started this investigation, is that when I converted the **trained**  onnx model to another runtime inference framework such as ncnn, the inference time is around 100ms, which is expected (in my experience, on my hardware, using 8 threads, it should take 50ms for r100 and 100ms for r200). I'm therefore wondering why onnxruntime seems to be much slower compared to this other framework. 

I have also run the above experiment with an R100 model, which has even more significant differences in inference time:
**ORT_ENABLE_EXTENDED**:
- R100 dummy avg inference time: 53.90166759490967 ms
- R100 trained avg inference time: 103.88010501861572 ms

**ORT_ENABLE_ALL**:
- R100 dummy avg inference time: 39.14788246154785 ms
- R100 trained avg inference time: 220.45928478240967 ms
 
So in the above example, I once again get the "expected" inference time of 50ms with the only initialized model, but for the trained model, it is twice as slow! If I run inference with the trained model using ncnn, I do indeed get 50ms. I therefore know that something strange is going on with onnxruntime. 

Can anyone let me know what is going on here? Ultimately what I want to know is how can I speed up inference of my trained models to match the inference speed of other inference frameworks. I would expected the onnxruntime CPU inference time to be competitive with, and even outperform, other frameworks such as ncnn, and not expect a 1.5 - 2x slowdown like I have demonstrated above.

The r100 model files can be downloaded here: 
- [r200_trained.onnx](https://drive.google.com/file/d/1MB8qr0tRAqQZxkI5I7DglbZ04gvK2dtp/view?usp=share_link)
- [r200_dummy.onnx](https://drive.google.com/file/d/1DlNg4DhhAU4XB8xtxJEi_yONaVnXBHt5/view?usp=share_link)

### To reproduce

See above

### Urgency

Somewhat higher than normal. Am in the process of swapping out our CPU inference framework to onnxruntime in a commercial product but am experiencing sudden performance slowdowns with many of our models, as described above. 

### Platform

Linux

### OS Version

20.04.5 LTS

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

1.14.1

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

### Model File

Please see above google drive links. 

### Is this a quantized model?

No

**Comment by cyrusbehr** [comment]
Just to add to my question some more evidence for my question. I converted the `r200_trained.onnx` model to ncnn using this tool [here](https://github.com/Tencent/ncnn/blob/master/tools/onnx/onnx2ncnn.cpp) using ncnn project commit: `5e4ea0b`. The resulting model file can be found [here](https://drive.google.com/file/d/1mIqfmcs90p_Mk43ftRJIYqZoYIJosuZC/view?usp=share_link).

Using the below c++ ncnn inference code, on the same hardware used to run the above benchmarks, with the same 8 threads, the inference time is 94ms, compared to the same 158ms when using onnxruntime. 

```c++
#include <iostream>
#include <opencv2/opencv.hpp>
#include "net.h"
#include <chrono>
typedef std::chrono::high_resolution_clock Clock;

int main() {
    // Load the image for inference
    auto img = cv::imread("../img/chip.jpg");
    if (img.empty()) {
        throw std::runtime_error("unable to read image at provided path");
    }

    cv::cvtColor(img, img, cv::COLOR_BGR2RGB);

    if (img.cols != 112 || img.rows != 112) {
        throw std::runtime_error("The image is of the wrong dims");
    }

    ncnn::Net net;
    net.load_param("../models/r200_trained/ncnn.param");
    net.load_model("../models/r200_trained/ncnn.bin");

    // Run 5 times to warm things up
    for (int i = 0; i < 5; ++i) {
        auto ex = net.create_extractor();
        ex.set_num_threads(8);

        ncnn::Mat in = ncnn::Mat::from_pixels(img.data, ncnn::Mat::PIXEL_RGB, 112, 112);
        const float meanVals[3] = {127.5, 127.5, 127.5};
        const float normVals[3] = {1 / 127.5, 1 / 127.5, 1 / 127.5};
        in.substract_mean_normalize(meanVals, normVals);

        ex.input("input", in);
        ncnn::Mat out;
        ex.extract("output", out);
    }

    size_t numIts = 100;
    auto start = Clock::now();
    for (size_t i = 0; i < numIts; ++i) {
        auto ex = net.create_extractor();
        ex.set_num_threads(8);
        ex.set_light_mode(true);

        ncnn::Mat in = ncnn::Mat::from_pixels(img.data, ncnn::Mat::PIXEL_RGB, 112, 112);
        const float meanVals[3] = {127.5, 127.5, 127.5};
        const float normVals[3] = {1 / 127.5, 1 / 127.5, 1 / 127.5};
        in.substract_mean_normalize(meanVals, normVals);

        ex.input("input", in);
        ncnn::Mat out;
        ex.extract("output", out);
    }
    auto end = Clock::now();
    double avgTime = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / static_cast<double>(numIts);
    std::cout << "Avg time: " << avgTime << "ms" << std::endl;

}

```

**Comment by xadupre** [comment]
Here is a snapshot of what the optimization is doing in your case on the beginning of your model (right side is optimized). I need to profile both models to check which added operator is causing the performance issue.

![image](https://user-images.githubusercontent.com/22452781/234061938-91d461bc-9f09-4ae2-bed8-3c85db6a841b.png)

**Comment by cyrusbehr** [comment]
Interesting @xadupre , I look forward to seeing the results of your analysis. 
Tomorrow, I will also provide you with the r100 models I mention above, in which the inference time for one model is 2-4x slower than the other.

**Comment by yf711** [comment]
Quick question @cyrusbehr, have you experienced regression when setting ORT_ENABLE_BASIC as optimization level?

On my side the regression issue is verified on all optimization levels except the basic level, which showed 20% gain compared to DISABLE_ALL

**Comment by xadupre** [comment]
Here are some results.

First graph shows that the issue comes operator Conv. The optimized has more Conv operator and they take much longer.

![plot_profiling](https://user-images.githubusercontent.com/22452781/234355356-687ee1d0-4424-44d0-bed2-be786ff0d70d.png)

Second graph shows that for the same operation Conv (same type and shape for the input and output), it takes much longer in the optimized model. The reason is they don't use the same implementation. The non-optimized model is using Conv from **ai.onnx**, the optimized model is using operator Conv from domain **com.microsoft.nchwc**.

The implementation is probably optimized for a different and the optimizer should not change the graph or the implementation is slower and needs to be fixed.

![plot_profiling_side_by_side](https://user-images.githubusercontent.com/22452781/234356081-cff98c7f-2ef1-43d5-b35e-7822517c6fd0.png)

**Comment by cyrusbehr** [comment]
@yf711 even setting the opt level to `ORT_ENABLE_BASIC` I still get: 
```
R200 dummy avg inference time: 104.35541868209839 ms
R200 trained avg inference time: 159.80592012405396 ms
``` 

@xadupre thank you for the above breakdown. Based on what you observed, what are you recommendations? 

Here is another even more extreme example (perhaps I should have started with this example). I have these two r100 models, exact same architecture, both trained. One of them takes 2x as long as the other for inference. I have no idea why. 

Only difference I see is in the onnx version: 
![image](https://user-images.githubusercontent.com/17056751/234403814-eaefafcc-04fe-4e4b-88d9-4d9a81d22177.png)


I used the following benchmark script to cycle through all the optimizations:
```python
import onnxruntime as ort
import numpy as np
from time import time


def benchmark_model(model_path, opt_level):
    sess_opt = ort.SessionOptions()
    sess_opt.intra_op_num_threads = 8
    sess_opt.graph_optimization_level = opt_level

    ort_session = ort.InferenceSession(model_path, sess_opt, providers=['CPUExecutionProvider'])
    output_names = [ort_session.get_outputs()[0].name]
    input_name = ort_session.get_inputs()[0].name

    # Run on dummy input
    img = np.ones((1,3,112,112)).astype('float32')

    # Run 5 times to warm things up
    for i in range(5):
        tp = ort_session.run(None, {input_name:img})

    # Benchmark
    num_iterations = 100
    start = time()
    for i in range(num_iterations):
        op = ort_session.run(None, {input_name:img})

    end = time()
    total_time = (end-start)/num_iterations * 1000
    print("Model:", model_path, ", time:", total_time)


opt_levels = [ort.GraphOptimizationLevel.ORT_DISABLE_ALL, ort.GraphOptimizationLevel.ORT_ENABLE_BASIC, 
ort.GraphOptimizationLevel.ORT_ENABLE_EXTENDED, ort.GraphOptimizationLevel.ORT_ENABLE_ALL]

for opt_level in opt_levels:
    print("Optimization level:", opt_level)
    benchmark_model("./r100_v1.onnx", opt_level)
    benchmark_model("./r100_v2.onnx", opt_level)
    print("")
```

Here are the results on the same hardware as above:
```
Optimization level: GraphOptimizationLevel.ORT_DISABLE_ALL
Model: ./r100_v1.onnx , time: 111.38367891311646
Model: ./r100_v2.onnx , time: 62.790102958679206

Optimization level: GraphOptimizationLevel.ORT_ENABLE_BASIC
Model: ./r100_v1.onnx , time: 117.80040979385376
Model: ./r100_v2.onnx , time: 65.80243587493896

Optimization level: GraphOptimizationLevel.ORT_ENABLE_EXTENDED
Model: ./r100_v1.onnx , time: 121.04552745819092
Model: ./r100_v2.onnx , time: 66.17196321487427

Optimization level: GraphOptimizationLevel.ORT_ENABLE_ALL
Model: ./r100_v1.onnx , time: 250.7680940628052
Model: ./r100_v2.onnx , time: 41.043078899383545
```
As can be seen, v1 is 2-4x slower than v2! @xadupre are you able to look into this as well please. Is this a bug in onnxruntime? Or what exactly is going on. With a different framework such as ncnn, both take ~40ms to run inference. 

[v1 download](https://drive.google.com/file/d/1t8_YadRW9-oLF1qp_-rYBcRSQYMCmAzi/view?usp=share_link)
[v2 download](https://drive.google.com/file/d/1OLA2ZryMBL7WinkPEEU2qjFCzrQyYcBX/view?usp=share_link)

**Comment by cyrusbehr** [comment]
And if the onnx version / ai.onnx version difference is indeed causing the slowdown, is there any way I can upgrade the onnx / ai.onnx version to the latest for an existing model without retraining?

**Comment by xadupre** [comment]
About the two new models you shared, v1 and v2 and two models similar or coming from different training? The fact the optimizations produce completely different results is interesting. For the time being, I suggest selecting the faster optimization parameter as @yf711 suggests. I'll see if I can determine a rule which makes one fast and the other one slow and then a fix.

**Comment by cyrusbehr** [comment]
I am told by our ML team that both v1 and v2 models used the same training script (if any changes were made to the script, they were very minor). v1 was trained a while back and is epoch 17 of our training, while v2 was trained this morning and epoch 1. The version of pytorch used to train the models was the same, and we exported both models to use the same onnx opset.

**Comment by cyrusbehr** [comment]
@xadupre I also had a coworker run the same test as me using the v1 and v2 models for the `ORT_ENABLE_ALL` optimization level, limiting to 8 threads, but instead of running on an intel CPU, he ran on an AMD EPYC 7502, and he got the same inference time for both v1 and v2 models (57ms). So seems this issue may only be present on intel CPUs.

**Comment by cyrusbehr** [comment]
@xadpure did you have a chance to look into the discrepancy between the two r100 models? This is somewhat urgent for my commercial deployment.

**Comment by xadupre** [comment]
Some intermediate results. All non-optimized models are using the same code, all optimized models are also using the same code. I looked into one particular configuration: Conv(X:[1, 256, 14, 14], W: [256, 256, 3, 3], B[256]). This is the duration of the execution of every node like this for 6 iterations. There is a pattern and the only difference are the coefficients.

![image](https://user-images.githubusercontent.com/22452781/235994411-af328de3-f0dc-4419-8386-6b200bfbb1fb.png)

So I made a small onnx graph with only one Conv operator with the same setting Conv(X:[1, 256, 14, 14], W: [256, 256, 3, 3], B[256]) where X is constant equal to 1, W is either drawn with a random law or the same weights as Conv_166 from r100_v1.onnx. B is null. I multiplied the coefficients by a factor called *scale*.

![image](https://user-images.githubusercontent.com/22452781/235995962-e923a8c0-5cb6-4cd4-ac01-0d309c8e8a63.png)

So there is something going on. I need to check if this behaviour is the same with the optimized implementation (based on your results, it is very likely), compare with torch, gpu, or with other types (float16).

Below, the slow model with just one Conv operator using one of the worse W matrix from r100_v1.onnx.

[slow_conv.zip](https://github.com/microsoft/onnxruntime/files/11388184/slow_conv.zip)

Appendix:

The same graph for model r100_v2.onnx:

![image](https://user-images.githubusercontent.com/22452781/235996862-601723c1-314f-43e6-a259-b40178a44e54.png)

The distribution of W coefficients in r100_v1.onnx:

![image](https://user-images.githubusercontent.com/22452781/235997402-de8bbbc2-43bf-4b45-86fb-d5faf976d893.png)

Same with r100_v2.onnx:

![image](https://user-images.githubusercontent.com/22452781/235997553-73b57858-fc38-4bd4-b354-f7c17e023478.png)

**Comment by cyrusbehr** [comment]
That is very interesting, thank you for the update.

**Comment by xadupre** [comment]
I did the same with torch and cuda and it gives the same results. I still need to figure out why exactly. In that graph, I took the model linked above and I measure the computation time of Conv(X, W * scale, B=0). scale is the X-axis.

![image](https://user-images.githubusercontent.com/22452781/236265102-9b6159f0-d62b-4130-8100-18ca7aa197e3.png)

**Comment by xadupre** [comment]
Thanks, it works.

Does it have an impact on GPU as well?

Xavier


Sent from Outlook<http://aka.ms/weboutlook>

________________________________
De : Yufeng Li ***@***.***>
Envoyé : jeudi 4 mai 2023 19:06
À : microsoft/onnxruntime ***@***.***>
Cc : Xavier Dupre ***@***.***>; Mention ***@***.***>
Objet : Re: [microsoft/onnxruntime] [Performance] Inference speed discrepancy between two r200 models on CPU (Issue #15630)


I think this is caused by subnormal values. Could you please try enabling flushing subnormal as zero, like this:

https://github.com/microsoft/onnxruntime/blob/8e610f25d8a14d6a4f311106d15d90d4f85aef91/onnxruntime/test/perftest/ort_test_session.cc#L769

**Comment by yufenglee** [comment]
No, it doesn't. only for CPU.
